Fixed problem in `CollectionDeserializer` where `_nullProvider` was not used if it was not `JsonToken.VALUE_NULL`

k163377 opened the pull request:
SSIA

fix for https://github.com/FasterXML/jackson-databind/issues/5139

It was a bit complicated, so I apologize if there are any problems with the modifications.

k163377 committed:
Fixed to call getNullValue if the deserialization result is null

Fixes https://github.com/FasterXML/jackson-databind/issues/5139

k163377 committed:
Add test cases

for https://github.com/FasterXML/jackson-databind/issues/5139

k163377 commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -481,6 +487,22 @@ protected Collection<Object> _deserializeWithObjectId(JsonParser p, Deserializat
+    ) throws IOException {
+        return _valueTypeDeserializer == null
+                ? _valueDeserializer.deserialize(p, ctxt)
+                : _valueDeserializer.deserializeWithType(p, ctxt, _valueTypeDeserializer);
```
`_valueTypeDeserializer` and `_valueDeserializer` are `final` properties and were also being warned by `Intellij`, so they have been changed to refer to them directly.

k163377 commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -444,25 +448,27 @@ protected Collection<Object> _deserializeWithObjectId(JsonParser p, Deserializat
+                }
+
+                if (value == null) {
+                    value = _nullProvider.getNullValue(ctxt);
```
Is it appropriate to call `_nullProvider` on a deserialized result using `_valueDeserializer`?

cowtowncoder replied:
I don't think so; it should be applied to input `null`s. Although TBH, value deserializers are also not expected to be handed `null` tokens ever, hence should rarely return `null`s too.

So no should not call on deserialized values.

**EDIT**: see my next comment, changed my mind.

cowtowncoder replied:
Actually, right. Based on fact `null` comes via coercion, I guess this must be done. No real way around it: null check/handling needed in both cases (content/input null, deserializer-provided (coerced) null)

k163377 committed:
Fix for compatibility with Java 8

k163377 commented on src/test/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer5139Test.java:
Is the name, placement and content of the test appropriate?
The following test was referenced for name and placement.
https://github.com/JooHyukKim/jackson-databind/commit/51a92b9359144d9288e52c942e3d576edff67ada

cowtowncoder replied:
Should go next to `CollectionDeserTest` in `com.fasterxml.jackson.databind.deser.jdk`. Name is fine.

k163377 replied:
Fixed
e4941a5

k163377 committed:
Move test

wrt https://github.com/FasterXML/jackson-databind/pull/5140#discussion_r2072432033

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer5139Test.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    public void nullsFailTest() {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.configOverride(List.class).setSetterInfo(JsonSetter.Value.forContentNulls(Nulls.FAIL));
```
Let's use `JsonMapper.builder()` approach so it merges more easily into 3.x (3.x has immutable `ObjectMapper`). 

Something like

```
mapper = JsonMapper.builder()
    .defaultSetterInfo(JsonSetter.Value.forContentNulls(Nulls.FAIL))
    .build();
```

k163377 replied:
Fixed
93e1a21

k163377 replied:
Oh, I am sorry.
There was a mistake and I have made an additional correction.
88bc7e4

cowtowncoder replied:
np, thanks!

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer5139Test.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Test
+    public void nullsSkipTest() throws JsonProcessingException {
```
also: use "throws Exception" instead (3.x has unchecked exceptions, renamed/removed `JsonProcessingException`)

k163377 replied:
Fixed
b8a3559

k163377 committed:
Fix to use JsonMapper.builder

wrt https://github.com/FasterXML/jackson-databind/pull/5140#discussion_r2072432619

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -341,12 +341,10 @@ protected Collection<Object> _deserializeFromArray(JsonParser p, Deserialization
         // [databind#631]: Assign current value, to be accessible by custom serializers
         p.assignCurrentValue(result);
 
-        JsonDeserializer<Object> valueDes = _valueDeserializer;
```
If it's ok, I'll merge these into 2.19 separately to reduce diff? And can then merge to PR.

cowtowncoder replied:
Done via #5141, 2.19 updated.

cowtowncoder replied:
Merged into PR to reduce diff.

k163377 committed:
Fix throws generic Exception

wrt https://github.com/FasterXML/jackson-databind/pull/5140#discussion_r2072432765

k163377 committed:
Fix copy miss

wrt 93e1a21cf9721bfacebfee0376fc01712b8cba88

cowtowncoder committed:
Merge branch '2.19' into fix/kotlin-976

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -353,12 +353,16 @@ protected Collection<Object> _deserializeFromArray(JsonParser p, Deserialization
                 }
+
+                // _skipNullValues is checked by _tryToAddNull.
                 if (value == null) {
```
This part can be moved within previous block, after `value = _nullProvider.getNullValue(ctxt);`

k163377 replied:
Fixed
aa4dbab

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -406,12 +411,15 @@ protected final Collection<Object> handleNonArray(JsonParser p, DeserializationC
+                value = _deserializeRawContentValue(p, ctxt);
+            }
+
+            if (value == null) {
```
Same here, can move to inside block

k163377 replied:
Fixed
aa4dbab

k163377 committed:
Reduce the number of times value == null is determined

wrt https://github.com/FasterXML/jackson-databind/pull/5140#discussion_r2072436774

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java:
```diff
@@ -475,6 +488,22 @@ protected Collection<Object> _deserializeWithObjectId(JsonParser p, Deserializat
+     * This method only performs deserialization and does not consider _skipNullValues, _nullProvider, etc.
+     * @since 2.19.1
+     */
+    protected Object _deserializeRawContentValue(
```
Method makes sense but I think I'll rename it and can then probably merge

cowtowncoder committed:
Update release notes, minor stylistic tweaks

cowtowncoder commented:
Excellent work as usual. Thank you @k163377 -- merged in for 2.19(.1), will merge to 3.x

Bit too risky to backport in 2.18, unfortunately, I think.

k163377 commented:
Thank you very much.

DavidRigglemanININ commented:
Is there an estimate of when 2.19.1 might be released? We are eagerly awaiting for this fix to be released. Thanks!

cowtowncoder commented:
@DavidRigglemanININ Unfortunately due to effort involved (3-4 hours for full patch release) it'll probably be couple of weeks yet. I need to find time to do that (not being part of my day job).
Usually it's a function of number of accumulated fixes, we have just 3 so far.

JooHyukKim commented:
Yes, with lots going on at the moment, you might want to use and overwrite with latest jar? (not an offical release, but with fix you need. Using Gradle or Maven I guess

JooHyukKim commented:
Yes, with lots going on at the moment, you might want to use and overwrite with latest jar? (not an offical release, but with fix you need. Using Gradle or Maven I guess. --In case you re urgent cc @DavidRigglemanININ .

DavidRigglemanININ commented:
I can wait a few weeks. It's not that urgent as we've been living for the pain for quite a while now (as we couldn't use the strictNullChecks on the kotlin module side due to performance issues). It's just been a pain point for us over the years that I'm really excited to finally get behind us as you can probably tell :). Thanks for all your hard work on this project!

cowtowncoder commented:
@DavidRigglemanININ Thank you! With Sonatype's aggressive push to shut down OSSRH I may end up actually making release earlier, although first needing to spend a bit of time on migrating publishing to use Sonatype Central Portal. We'll see.

DavidRigglemanININ commented:
> @DavidRigglemanININ Thank you! With Sonatype's aggressive push to shut down OSSRH I may end up actually making release earlier, although first needing to spend a bit of time on migrating publishing to use Sonatype Central Portal. We'll see.

Are we any closer to a release in the next few weeks?

cowtowncoder commented:
@DavidRigglemanININ I think so -- I do want to get 2.19.1 out before I go on my vacation in about 2 weeks.
